pyop2/global_kernel.py

```diff
@@ -336,6 +389,103 @@ def code_to_compile(self):
             return preamble + "\nextern \"C\" {\n" + device_code + "\n}\n"
         return code.device_code()
 
+    def vectorise(self, wrapper, iname, batch_size):
+        """Return a vectorised version of wrapper, vectorising over iname.
+
+        :arg wrapper: A loopy kernel to vectorise.
+        :arg iname: The iteration index to vectorise over.
+        :arg batch_size: The vector width."""
+        if batch_size == 1:
+            return wrapper
+
+        from functools import reduce
+        import pymbolic.primitives as prim
+
+        if not configuration["debug"]:
+            # loopy warns for every instruction that cannot be vectorized;
+            # ignore in non-debug mode.
+            new_entrypoint = wrapper.default_entrypoint.copy(
+                silenced_warnings=(wrapper.default_entrypoint.silenced_warnings
+                                   + ["vectorize_failed"]))
+            wrapper = wrapper.with_kernel(new_entrypoint)
+
+        kernel = wrapper.default_entrypoint
+
+        # align temps
+        alignment = configuration["alignment"]
+        tmps = {name: tv.copy(alignment=alignment)
+                for name, tv in kernel.temporary_variables.items()}
+        kernel = kernel.copy(temporary_variables=tmps)
+
+        # {{{ record temps that cannot be vectorized
+
+        # Do not vectorize temporaries used outside *iname*
+        temps_not_to_vectorize = reduce(set.union,
+                                        [(insn.dependency_names()
+                                          & frozenset(kernel.temporary_variables))
+                                         for insn in kernel.instructions
+                                         if iname not in insn.within_inames],
+                                        set())
+
+        # Constant literal temporaries are arguments => cannot vectorize
+        temps_not_to_vectorize |= {name
+                                   for name, tv in kernel.temporary_variables.items()
+                                   if (tv.read_only
+                                       and tv.initializer is not None)}
+
+        # {{{ clang (unlike gcc) does not allow taking address of vector-type
+        # variable
+
+        # FIXME: Perform this only if we know we are not using gcc.
+        for insn in kernel.instructions:
```

Comment:
This seems to be still failing if compiled with clang. Adding the temps in the call to `temps_not_to_vectorize` might not be sufficient.

Reply:
Thanks, fixed a typo in [a671b6c](https://github.com/OP2/PyOP2/pull/654/commits/a671b6ccb50e2ed7b12df1800534f99979bdb9fb). Looked at the generated code, looks fine now.